File: 1167_minimum_cost_to_connect_sticks/connect_sticks.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

/**
 * Time: O(n * log(n))
 * Space: O(n)
 */
int connectSticks(vector<int> &sticks)
{
    // Syntax to create a min heap for priority queue
    // priority_queue <int, vector<int>, greater<int>> g = gq;
    priority_queue<int, vector<int>, greater<int>> qu(sticks.begin(), sticks.end());
    int result = 0;
    while (qu.size() > 1)
    {
        int a = qu.top();
        qu.pop();
        int b = qu.top();
        qu.pop();
        result += a + b;
        qu.push(a + b);
    }
    return result;
}

File: 1167_minimum_cost_to_connect_sticks/connect_sticks.py
import heapq
from typing import List


def connect_sticks(sticks: List[int]) -> int:
    """
    Time: O(n * log(n))
    Space: O(n)
    """
    heapq.heapify(sticks)
    res = 0
    while len(sticks) > 1:
        f, s = heapq.heappop(sticks), heapq.heappop(sticks)
        tmp = f + s
        res += tmp
        heapq.heappush(sticks, tmp)
    return res
